Exclude DAV CORS handling for non-browsers

PVince81 opened the pull request:
## Description
CORS only makes sense for browsers and davfs has trouble with the OPTIONS logic from CORS.

## Related Issue
Fixes https://github.com/owncloud/core/issues/29793

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?
- [x] TEST: Manual local test with davfs.
- [x] TEST: rerun sample app tests

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

DeepDiver1975 commented on apps/dav/lib/Connector/Sabre/CorsPlugin.php:
```diff
@@ -66,6 +67,12 @@ public function __construct(\OCP\IUserSession $userSession) {
 	public function initialize(\Sabre\DAV\Server $server) {
 		$this->server = $server;
 
+		if (strpos(strtolower($this->server->httpRequest->getHeader('user-agent')), 'mozilla/') === false) {
```
Matching on the user agent is not the right thing to do as far as I can tell after reading a bit into the specs
If I understand correctly we shall only apply CORS handling on the server if the 'Origin' header is sent with the request ...

see 
https://www.html5rocks.com/en/tutorials/cors/
https://spring.io/understanding/CORS
https://github.com/nelmio/NelmioCorsBundle/blob/master/EventListener/CorsListener.php#L68

PVince81 replied:
That's even better, I'll adjust accordingly. Thanks for pointing out.

PVince81 commented:
I've adjusted and squashed to use the "Origin" header only for detection.

Tested and works fine with davfs mounting and also the example app from https://github.com/noveens/staticOwncloudWebapp

@DeepDiver1975

PVince81 commented on apps/dav/lib/Connector/Sabre/CorsPlugin.php:
```diff
@@ -49,6 +49,7 @@ class CorsPlugin extends \Sabre\DAV\ServerPlugin {
 	public function __construct(\OCP\IUserSession $userSession) {
 		$this->userSession = $userSession;
 		$this->extraHeaders['Access-Control-Allow-Headers'] = ["X-OC-Mtime", "OC-Checksum", "OC-Total-Length", "Depth", "Destination", "Overwrite"];
+		// TODO: use $this->server->getAllowedMethods($request->getPath()) in the right location
```
https://github.com/owncloud/core/pull/29813

PVince81 commented:
- [ ] check if Origin is the same as current URL and skip CORS if that's the same

PVince81 commented:
I pushed something, unfinished... Need a utility function to properly compare domain + port with defaults (80, 443) in case they are specified still in case of quirky clients.

Commit:
Add utility functions to compare domains

Normalizes the domains and compares them based on protocol, host and
port.

Commit:
Exclude DAV CORS handling when no Origin specified

This will exclude non-browser clients from CORS handling.

Fixes some clients like davfs which break when CORS is enabled.

PVince81 commented:
Rebased, squashed.

I've added a `\OCP\Util::isSameDomain($url1, $url2)` function now in the first commit, with unit tests.
Then this method is used to check by CorsPlugin to detect cross-domain situations. If not cross-domain, it will skip CORS.

@DeepDiver1975 please re-review.

PVince81 commented on lib/public/Util.php:
```diff
@@ -749,4 +749,60 @@ public static function getStatusInfo($includeVersion = false, $serverHide = fals
+	/**
+	 * Check whether the given URLs have the same protocol, domain and port.
+	 * This is useful to check a browser's cross-domain situation.
+	 * If this method returns true, a browser would consider both URLs to be
```
changing this to `false`

DeepDiver1975 commented on lib/public/Util.php:
```diff
@@ -722,6 +722,7 @@ public static function getStatusInfo($includeVersion = false, $serverHide = fals
 
 		$installed = (bool) $systemConfig->getValue('installed', false);
 		$maintenance = (bool) $systemConfig->getValue('maintenance', false);
+		\OC_App::loadApps(['theme']);
```
seems unrelated - rebase left over?

PVince81 replied:
indeed, no idea how this got here.

DeepDiver1975 replied:
no worries - stuff happens

PVince81 commented:
@DeepDiver1975 I've removed the bad line.

PVince81 commented:
stable10: https://github.com/owncloud/core/pull/29874
